<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Subsetting PyRanges</title>
  <meta name="description" content="Subsetting PyRanges">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Subsetting PyRanges" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Subsetting PyRanges" />
  
  
  

<meta name="author" content="Endre Bakken Stovner">


<meta name="date" content="2022-01-11">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="using-multiple-cores.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction to PyRanges</a></li>
<li class="chapter" data-level="2" data-path="loadingcreating-pyranges.html"><a href="loadingcreating-pyranges.html"><i class="fa fa-check"></i><b>2</b> Loading/Creating PyRanges</a></li>
<li class="chapter" data-level="3" data-path="writing-pyranges-to-disk.html"><a href="writing-pyranges-to-disk.html"><i class="fa fa-check"></i><b>3</b> Writing PyRanges to disk</a></li>
<li class="chapter" data-level="4" data-path="subsetting-pyranges.html"><a href="subsetting-pyranges.html"><i class="fa fa-check"></i><b>4</b> Subsetting PyRanges</a></li>
<li class="chapter" data-level="5" data-path="manipulating-the-data-in-pyranges.html"><a href="manipulating-the-data-in-pyranges.html"><i class="fa fa-check"></i><b>5</b> Manipulating the data in PyRanges</a></li>
<li class="chapter" data-level="6" data-path="concatenating-pyranges.html"><a href="concatenating-pyranges.html"><i class="fa fa-check"></i><b>6</b> Concatenating PyRanges</a></li>
<li class="chapter" data-level="7" data-path="piping-pyranges.html"><a href="piping-pyranges.html"><i class="fa fa-check"></i><b>7</b> Piping PyRanges</a></li>
<li class="chapter" data-level="8" data-path="printing-pyranges.html"><a href="printing-pyranges.html"><i class="fa fa-check"></i><b>8</b> Printing PyRanges</a></li>
<li class="chapter" data-level="9" data-path="iterating-over-pyranges.html"><a href="iterating-over-pyranges.html"><i class="fa fa-check"></i><b>9</b> Iterating over PyRanges</a></li>
<li class="chapter" data-level="10" data-path="sorting-pyranges.html"><a href="sorting-pyranges.html"><i class="fa fa-check"></i><b>10</b> Sorting PyRanges</a></li>
<li class="chapter" data-level="11" data-path="summarizing-pyranges.html"><a href="summarizing-pyranges.html"><i class="fa fa-check"></i><b>11</b> Summarizing PyRanges</a></li>
<li class="chapter" data-level="12" data-path="methods-for-manipulating-single-pyranges.html"><a href="methods-for-manipulating-single-pyranges.html"><i class="fa fa-check"></i><b>12</b> Methods for manipulating single PyRanges</a></li>
<li class="chapter" data-level="13" data-path="applying-custom-methods-to-pairs-of-pyranges.html"><a href="applying-custom-methods-to-pairs-of-pyranges.html"><i class="fa fa-check"></i><b>13</b> Applying custom methods to pairs of PyRanges</a></li>
<li class="chapter" data-level="14" data-path="intersecting-ranges.html"><a href="intersecting-ranges.html"><i class="fa fa-check"></i><b>14</b> Intersecting Ranges</a></li>
<li class="chapter" data-level="15" data-path="overlapping-ranges.html"><a href="overlapping-ranges.html"><i class="fa fa-check"></i><b>15</b> Overlapping Ranges</a></li>
<li class="chapter" data-level="16" data-path="joining-ranges.html"><a href="joining-ranges.html"><i class="fa fa-check"></i><b>16</b> Joining Ranges</a></li>
<li class="chapter" data-level="17" data-path="finding-the-closest-intervals.html"><a href="finding-the-closest-intervals.html"><i class="fa fa-check"></i><b>17</b> Finding the closest intervals</a></li>
<li class="chapter" data-level="18" data-path="finding-the-k-closest-intervals.html"><a href="finding-the-k-closest-intervals.html"><i class="fa fa-check"></i><b>18</b> Finding the k closest intervals</a></li>
<li class="chapter" data-level="19" data-path="create-count-matrix-from-multiple-pyranges.html"><a href="create-count-matrix-from-multiple-pyranges.html"><i class="fa fa-check"></i><b>19</b> Create count-matrix from multiple PyRanges</a></li>
<li class="chapter" data-level="20" data-path="statistics-colocalization-and-co-occurence-measures.html"><a href="statistics-colocalization-and-co-occurence-measures.html"><i class="fa fa-check"></i><b>20</b> Statistics: colocalization and co-occurence measures</a></li>
<li class="chapter" data-level="21" data-path="statistics-simes-method.html"><a href="statistics-simes-method.html"><i class="fa fa-check"></i><b>21</b> Statistics: Simes method</a></li>
<li class="chapter" data-level="22" data-path="statistics-compute-fisher-exact.html"><a href="statistics-compute-fisher-exact.html"><i class="fa fa-check"></i><b>22</b> Statistics: compute Fisher exact</a></li>
<li class="chapter" data-level="23" data-path="statistics-computing-the-matthews-correlation-coeficcient-between-ranges.html"><a href="statistics-computing-the-matthews-correlation-coeficcient-between-ranges.html"><i class="fa fa-check"></i><b>23</b> Statistics: computing the Matthews correlation coeficcient between ranges</a></li>
<li class="chapter" data-level="24" data-path="statistics-rowbased-operations.html"><a href="statistics-rowbased-operations.html"><i class="fa fa-check"></i><b>24</b> Statistics: rowbased operations</a></li>
<li class="chapter" data-level="25" data-path="turning-ranges-into-rles.html"><a href="turning-ranges-into-rles.html"><i class="fa fa-check"></i><b>25</b> Turning Ranges into RLEs</a></li>
<li class="chapter" data-level="26" data-path="an-introduction-to-rles.html"><a href="an-introduction-to-rles.html"><i class="fa fa-check"></i><b>26</b> An introduction to RLEs</a></li>
<li class="chapter" data-level="27" data-path="run-length-encoding-dicts.html"><a href="run-length-encoding-dicts.html"><i class="fa fa-check"></i><b>27</b> Run Length Encoding dicts</a></li>
<li class="chapter" data-level="28" data-path="subsetting-rles.html"><a href="subsetting-rles.html"><i class="fa fa-check"></i><b>28</b> Subsetting Rles</a></li>
<li class="chapter" data-level="29" data-path="subsetting-pyrles.html"><a href="subsetting-pyrles.html"><i class="fa fa-check"></i><b>29</b> Subsetting PyRles</a></li>
<li class="chapter" data-level="30" data-path="using-multiple-cores.html"><a href="using-multiple-cores.html"><i class="fa fa-check"></i><b>30</b> Using multiple cores</a></li>
<li class="chapter" data-level="31" data-path="genomicfeatures-methods-using-genomic-context.html"><a href="genomicfeatures-methods-using-genomic-context.html"><i class="fa fa-check"></i><b>31</b> GenomicFeatures: methods using genomic context</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Subsetting PyRanges</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="genomicfeatures-methods-using-genomic-context" class="section level1" number="31">
<h1><span class="header-section-number">31</span> GenomicFeatures: methods using genomic context</h1>
<p>The PyRanges library has a few methods that operate on a genomic context. Their
functions are found in the <code>pyranges.gf</code> namespace and the methods to be used on
a PyRanges-object are found in the <code>gr.features</code> namespace.</p>
<p><code>introns</code> can be used on pyranges annotated with gene_id and exon_id.</p>
<div class="sourceCode" id="cb328"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb328-1"><a href="genomicfeatures-methods-using-genomic-context.html#cb328-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyranges <span class="im">as</span> pr</span>
<span id="cb328-2"><a href="genomicfeatures-methods-using-genomic-context.html#cb328-2" aria-hidden="true" tabindex="-1"></a>gr <span class="op">=</span> pr.data.ucsc_bed()</span>
<span id="cb328-3"><a href="genomicfeatures-methods-using-genomic-context.html#cb328-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gr)</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+------------+-------+
## | Chromosome   | Start     | End       | Feature    | gene_id    | +4    |
## | (category)   | (int32)   | (int32)   | (object)   | (object)   | ...   |
## |--------------+-----------+-----------+------------+------------+-------|
## | chr1         | 12776117  | 12788726  | gene       | AADACL3    | ...   |
## | chr1         | 169075927 | 169101957 | gene       | ATP1B1     | ...   |
## | chr1         | 6845383   | 7829766   | gene       | CAMTA1     | ...   |
## | chr1         | 20915589  | 20945396  | gene       | CDA        | ...   |
## | ...          | ...       | ...       | ...        | ...        | ...   |
## | chrX         | 152661096 | 152663330 | exon       | PNMA6E     | ...   |
## | chrX         | 152661096 | 152666808 | transcript | PNMA6E     | ...   |
## | chrX         | 152664164 | 152664378 | exon       | PNMA6E     | ...   |
## | chrX         | 152666701 | 152666808 | exon       | PNMA6E     | ...   |
## +--------------+-----------+-----------+------------+------------+-------+
## Stranded PyRanges object has 5,519 rows and 9 columns from 30 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 4 hidden columns: transcript_id, Strand, exon_number, transcript_name</code></pre>
<div class="sourceCode" id="cb330"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb330-1"><a href="genomicfeatures-methods-using-genomic-context.html#cb330-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gr.features.introns(by<span class="op">=</span><span class="st">&quot;transcript&quot;</span>)) <span class="co"># default is by gene</span></span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+--------------+-------+
## | Chromosome   | Start     | End       | Feature    | gene_id      | +4    |
## | (object)     | (int32)   | (int32)   | (object)   | (object)     | ...   |
## |--------------+-----------+-----------+------------+--------------+-------|
## | chr1         | 12227     | 12612     | intron     | LOC102725121 | ...   |
## | chr1         | 12721     | 13220     | intron     | LOC102725121 | ...   |
## | chr1         | 12227     | 12612     | intron     | DDX11L1      | ...   |
## | chr1         | 12721     | 13220     | intron     | DDX11L1      | ...   |
## | ...          | ...       | ...       | ...        | ...          | ...   |
## | chrX         | 9714193   | 9716613   | intron     | GPR143       | ...   |
## | chrX         | 9716706   | 9727371   | intron     | GPR143       | ...   |
## | chrX         | 9727466   | 9728756   | intron     | GPR143       | ...   |
## | chrX         | 9728866   | 9733607   | intron     | GPR143       | ...   |
## +--------------+-----------+-----------+------------+--------------+-------+
## Stranded PyRanges object has 4,128 rows and 9 columns from 30 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 4 hidden columns: transcript_id, Strand, exon_number, transcript_name</code></pre>
<p><code>tes/tss</code> can be used to find the starts and ends of transcripts:</p>
<div class="sourceCode" id="cb332"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb332-1"><a href="genomicfeatures-methods-using-genomic-context.html#cb332-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyranges <span class="im">as</span> pr</span>
<span id="cb332-2"><a href="genomicfeatures-methods-using-genomic-context.html#cb332-2" aria-hidden="true" tabindex="-1"></a>gr <span class="op">=</span> pr.data.ucsc_bed()</span>
<span id="cb332-3"><a href="genomicfeatures-methods-using-genomic-context.html#cb332-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gr[gr.Feature <span class="op">==</span> <span class="st">&quot;transcript&quot;</span>])</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+--------------+-------+
## | Chromosome   | Start     | End       | Feature    | gene_id      | +4    |
## | (category)   | (int32)   | (int32)   | (object)   | (object)     | ...   |
## |--------------+-----------+-----------+------------+--------------+-------|
## | chr1         | 11868     | 14362     | transcript | LOC102725121 | ...   |
## | chr1         | 11873     | 14409     | transcript | DDX11L1      | ...   |
## | chr1         | 30365     | 30503     | transcript | MIR1302-2    | ...   |
## | chr1         | 30365     | 30503     | transcript | MIR1302-9    | ...   |
## | ...          | ...       | ...       | ...        | ...          | ...   |
## | chrX         | 131337052 | 131352061 | transcript | RAP2C        | ...   |
## | chrX         | 134021661 | 134049287 | transcript | MOSPD1       | ...   |
## | chrX         | 152157367 | 152160757 | transcript | PNMA5        | ...   |
## | chrX         | 152661096 | 152666808 | transcript | PNMA6E       | ...   |
## +--------------+-----------+-----------+------------+--------------+-------+
## Stranded PyRanges object has 500 rows and 9 columns from 30 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 4 hidden columns: transcript_id, Strand, exon_number, transcript_name</code></pre>
<div class="sourceCode" id="cb334"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb334-1"><a href="genomicfeatures-methods-using-genomic-context.html#cb334-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gr.features.tes()) <span class="co"># default is by gene</span></span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+--------------+-------+
## | Chromosome   | Start     | End       | Feature    | gene_id      | +4    |
## | (category)   | (int32)   | (int32)   | (object)   | (object)     | ...   |
## |--------------+-----------+-----------+------------+--------------+-------|
## | chr1         | 14361     | 14362     | tes        | LOC102725121 | ...   |
## | chr1         | 14408     | 14409     | tes        | DDX11L1      | ...   |
## | chr1         | 30502     | 30503     | tes        | MIR1302-2    | ...   |
## | chr1         | 30502     | 30503     | tes        | MIR1302-9    | ...   |
## | ...          | ...       | ...       | ...        | ...          | ...   |
## | chrX         | 131337052 | 131337053 | tes        | RAP2C        | ...   |
## | chrX         | 134021661 | 134021662 | tes        | MOSPD1       | ...   |
## | chrX         | 152157367 | 152157368 | tes        | PNMA5        | ...   |
## | chrX         | 152661096 | 152661097 | tes        | PNMA6E       | ...   |
## +--------------+-----------+-----------+------------+--------------+-------+
## Stranded PyRanges object has 500 rows and 9 columns from 30 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 4 hidden columns: transcript_id, Strand, exon_number, transcript_name</code></pre>
<p><code>tile_genome</code> splits a PyRanges of chromosome sizes into a tiled genome.</p>
<div class="sourceCode" id="cb336"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb336-1"><a href="genomicfeatures-methods-using-genomic-context.html#cb336-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyranges <span class="im">as</span> pr</span>
<span id="cb336-2"><a href="genomicfeatures-methods-using-genomic-context.html#cb336-2" aria-hidden="true" tabindex="-1"></a>cs <span class="op">=</span> pr.data.chromsizes()</span>
<span id="cb336-3"><a href="genomicfeatures-methods-using-genomic-context.html#cb336-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cs)</span>
<span id="cb336-4"><a href="genomicfeatures-methods-using-genomic-context.html#cb336-4" aria-hidden="true" tabindex="-1"></a><span class="co"># can also do</span></span>
<span id="cb336-5"><a href="genomicfeatures-methods-using-genomic-context.html#cb336-5" aria-hidden="true" tabindex="-1"></a><span class="co"># pip install pyranges_db as db</span></span>
<span id="cb336-6"><a href="genomicfeatures-methods-using-genomic-context.html#cb336-6" aria-hidden="true" tabindex="-1"></a><span class="co"># cs =  db.ucsc.chromsizes(&quot;hg19&quot;)</span></span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+
## | Chromosome   | Start     | End       |
## | (category)   | (int32)   | (int32)   |
## |--------------+-----------+-----------|
## | chr1         | 0         | 249250621 |
## | chr2         | 0         | 243199373 |
## | chr3         | 0         | 198022430 |
## | chr4         | 0         | 191154276 |
## | ...          | ...       | ...       |
## | chr22        | 0         | 51304566  |
## | chrM         | 0         | 16571     |
## | chrX         | 0         | 155270560 |
## | chrY         | 0         | 59373566  |
## +--------------+-----------+-----------+
## Unstranded PyRanges object has 25 rows and 3 columns from 25 chromosomes.
## For printing, the PyRanges was sorted on Chromosome.</code></pre>
<div class="sourceCode" id="cb338"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb338-1"><a href="genomicfeatures-methods-using-genomic-context.html#cb338-1" aria-hidden="true" tabindex="-1"></a>tile_size <span class="op">=</span> <span class="bu">int</span>(<span class="fl">1e6</span>)</span>
<span id="cb338-2"><a href="genomicfeatures-methods-using-genomic-context.html#cb338-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(pr.gf.tile_genome(cs, tile_size, tile_last<span class="op">=</span><span class="va">False</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+
## | Chromosome   | Start     | End       |
## | (category)   | (int32)   | (int32)   |
## |--------------+-----------+-----------|
## | chr1         | 0         | 1000000   |
## | chr1         | 1000000   | 2000000   |
## | chr1         | 2000000   | 3000000   |
## | chr1         | 3000000   | 4000000   |
## | ...          | ...       | ...       |
## | chrY         | 56000000  | 57000000  |
## | chrY         | 57000000  | 58000000  |
## | chrY         | 58000000  | 59000000  |
## | chrY         | 59000000  | 59373566  |
## +--------------+-----------+-----------+
## Unstranded PyRanges object has 3,114 rows and 3 columns from 25 chromosomes.
## For printing, the PyRanges was sorted on Chromosome.</code></pre>
<div class="sourceCode" id="cb340"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb340-1"><a href="genomicfeatures-methods-using-genomic-context.html#cb340-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(pr.gf.tile_genome(cs, tile_size, tile_last<span class="op">=</span><span class="va">True</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+
## | Chromosome   | Start     | End       |
## | (category)   | (int32)   | (int32)   |
## |--------------+-----------+-----------|
## | chr1         | 0         | 1000000   |
## | chr1         | 1000000   | 2000000   |
## | chr1         | 2000000   | 3000000   |
## | chr1         | 3000000   | 4000000   |
## | ...          | ...       | ...       |
## | chrY         | 56000000  | 57000000  |
## | chrY         | 57000000  | 58000000  |
## | chrY         | 58000000  | 59000000  |
## | chrY         | 59000000  | 60000000  |
## +--------------+-----------+-----------+
## Unstranded PyRanges object has 3,114 rows and 3 columns from 25 chromosomes.
## For printing, the PyRanges was sorted on Chromosome.</code></pre>
<p><code>genome_bounds</code> removes all intervals in the PyRanges which are outside the
genome bounds. If the flag clip is used, the parts of the intervals which are
inside the boundaries are kept.</p>
<div class="sourceCode" id="cb342"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb342-1"><a href="genomicfeatures-methods-using-genomic-context.html#cb342-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyranges <span class="im">as</span> pr</span>
<span id="cb342-2"><a href="genomicfeatures-methods-using-genomic-context.html#cb342-2" aria-hidden="true" tabindex="-1"></a>cs <span class="op">=</span> pr.data.chromsizes()</span>
<span id="cb342-3"><a href="genomicfeatures-methods-using-genomic-context.html#cb342-3" aria-hidden="true" tabindex="-1"></a>gr <span class="op">=</span> pr.data.chipseq()</span>
<span id="cb342-4"><a href="genomicfeatures-methods-using-genomic-context.html#cb342-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cs)</span>
<span id="cb342-5"><a href="genomicfeatures-methods-using-genomic-context.html#cb342-5" aria-hidden="true" tabindex="-1"></a><span class="co"># can also do</span></span>
<span id="cb342-6"><a href="genomicfeatures-methods-using-genomic-context.html#cb342-6" aria-hidden="true" tabindex="-1"></a><span class="co"># pip install pyranges_db as db</span></span>
<span id="cb342-7"><a href="genomicfeatures-methods-using-genomic-context.html#cb342-7" aria-hidden="true" tabindex="-1"></a><span class="co"># cs =  db.ucsc.chromsizes(&quot;hg19&quot;)</span></span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+
## | Chromosome   | Start     | End       |
## | (category)   | (int32)   | (int32)   |
## |--------------+-----------+-----------|
## | chr1         | 0         | 249250621 |
## | chr2         | 0         | 243199373 |
## | chr3         | 0         | 198022430 |
## | chr4         | 0         | 191154276 |
## | ...          | ...       | ...       |
## | chr22        | 0         | 51304566  |
## | chrM         | 0         | 16571     |
## | chrX         | 0         | 155270560 |
## | chrY         | 0         | 59373566  |
## +--------------+-----------+-----------+
## Unstranded PyRanges object has 25 rows and 3 columns from 25 chromosomes.
## For printing, the PyRanges was sorted on Chromosome.</code></pre>
<div class="sourceCode" id="cb344"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb344-1"><a href="genomicfeatures-methods-using-genomic-context.html#cb344-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(pr.gf.genome_bounds(gr, cs, clip<span class="op">=</span><span class="va">True</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+--------------+
## | Chromosome   | Start     | End       | Name       | Score     | Strand       |
## | (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |
## |--------------+-----------+-----------+------------+-----------+--------------|
## | chr1         | 212609534 | 212609559 | U0         | 0         | +            |
## | chr1         | 169887529 | 169887554 | U0         | 0         | +            |
## | chr1         | 216711011 | 216711036 | U0         | 0         | +            |
## | chr1         | 144227079 | 144227104 | U0         | 0         | +            |
## | ...          | ...       | ...       | ...        | ...       | ...          |
## | chrY         | 15224235  | 15224260  | U0         | 0         | -            |
## | chrY         | 13517892  | 13517917  | U0         | 0         | -            |
## | chrY         | 8010951   | 8010976   | U0         | 0         | -            |
## | chrY         | 7405376   | 7405401   | U0         | 0         | -            |
## +--------------+-----------+-----------+------------+-----------+--------------+
## Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.</code></pre>
<div class="sourceCode" id="cb346"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb346-1"><a href="genomicfeatures-methods-using-genomic-context.html#cb346-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(pr.gf.genome_bounds(gr, cs))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+--------------+
## | Chromosome   | Start     | End       | Name       | Score     | Strand       |
## | (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |
## |--------------+-----------+-----------+------------+-----------+--------------|
## | chr1         | 212609534 | 212609559 | U0         | 0         | +            |
## | chr1         | 169887529 | 169887554 | U0         | 0         | +            |
## | chr1         | 216711011 | 216711036 | U0         | 0         | +            |
## | chr1         | 144227079 | 144227104 | U0         | 0         | +            |
## | ...          | ...       | ...       | ...        | ...       | ...          |
## | chrY         | 15224235  | 15224260  | U0         | 0         | -            |
## | chrY         | 13517892  | 13517917  | U0         | 0         | -            |
## | chrY         | 8010951   | 8010976   | U0         | 0         | -            |
## | chrY         | 7405376   | 7405401   | U0         | 0         | -            |
## +--------------+-----------+-----------+------------+-----------+--------------+
## Stranded PyRanges object has 9,979 rows and 6 columns from 24 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.</code></pre>
<p><code>random</code> creates a random PyRanges from the PyRanges of chromosome sizes given:</p>
<div class="sourceCode" id="cb348"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb348-1"><a href="genomicfeatures-methods-using-genomic-context.html#cb348-1" aria-hidden="true" tabindex="-1"></a>pr.random(n<span class="op">=</span><span class="dv">1000</span>, length<span class="op">=</span><span class="dv">100</span>, chromsizes<span class="op">=</span><span class="va">None</span>, strand<span class="op">=</span><span class="va">True</span>)</span></code></pre></div>
<p>if no chromsize is given, hg19 is used (from pr.data.chromsizes).</p>

</div>
<div class="footnotes footnotes-end-of-document">
<hr />
<ol>
<li id="fn1"><p>This is the same behavior as bedtools intersect.<a href="intersecting-ranges.html#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>This is the same behavior as Bioconductor GenomicRanges intersect.<a href="intersecting-ranges.html#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="using-multiple-cores.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
