<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>PyRanges documentation</title>
  <meta name="description" content="PyRanges documentation">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="PyRanges documentation" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="PyRanges documentation" />
  
  
  

<meta name="author" content="Endre Bakken Stovner">


<meta name="date" content="2022-01-12">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="overlapping-ranges.html">
<link rel="next" href="finding-the-closest-intervals.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction to PyRanges</a></li>
<li class="chapter" data-level="2" data-path="loadingcreating-pyranges.html"><a href="loadingcreating-pyranges.html"><i class="fa fa-check"></i><b>2</b> Loading/Creating PyRanges</a></li>
<li class="chapter" data-level="3" data-path="writing-pyranges-to-disk.html"><a href="writing-pyranges-to-disk.html"><i class="fa fa-check"></i><b>3</b> Writing PyRanges to disk</a></li>
<li class="chapter" data-level="4" data-path="subsetting-pyranges.html"><a href="subsetting-pyranges.html"><i class="fa fa-check"></i><b>4</b> Subsetting PyRanges</a></li>
<li class="chapter" data-level="5" data-path="manipulating-the-data-in-pyranges.html"><a href="manipulating-the-data-in-pyranges.html"><i class="fa fa-check"></i><b>5</b> Manipulating the data in PyRanges</a></li>
<li class="chapter" data-level="6" data-path="concatenating-pyranges.html"><a href="concatenating-pyranges.html"><i class="fa fa-check"></i><b>6</b> Concatenating PyRanges</a></li>
<li class="chapter" data-level="7" data-path="piping-pyranges.html"><a href="piping-pyranges.html"><i class="fa fa-check"></i><b>7</b> Piping PyRanges</a></li>
<li class="chapter" data-level="8" data-path="printing-pyranges.html"><a href="printing-pyranges.html"><i class="fa fa-check"></i><b>8</b> Printing PyRanges</a></li>
<li class="chapter" data-level="9" data-path="iterating-over-pyranges.html"><a href="iterating-over-pyranges.html"><i class="fa fa-check"></i><b>9</b> Iterating over PyRanges</a></li>
<li class="chapter" data-level="10" data-path="sorting-pyranges.html"><a href="sorting-pyranges.html"><i class="fa fa-check"></i><b>10</b> Sorting PyRanges</a></li>
<li class="chapter" data-level="11" data-path="summarizing-pyranges.html"><a href="summarizing-pyranges.html"><i class="fa fa-check"></i><b>11</b> Summarizing PyRanges</a></li>
<li class="chapter" data-level="12" data-path="methods-for-manipulating-single-pyranges.html"><a href="methods-for-manipulating-single-pyranges.html"><i class="fa fa-check"></i><b>12</b> Methods for manipulating single PyRanges</a></li>
<li class="chapter" data-level="13" data-path="applying-custom-methods-to-pairs-of-pyranges.html"><a href="applying-custom-methods-to-pairs-of-pyranges.html"><i class="fa fa-check"></i><b>13</b> Applying custom methods to pairs of PyRanges</a></li>
<li class="chapter" data-level="14" data-path="intersecting-ranges.html"><a href="intersecting-ranges.html"><i class="fa fa-check"></i><b>14</b> Intersecting Ranges</a></li>
<li class="chapter" data-level="15" data-path="overlapping-ranges.html"><a href="overlapping-ranges.html"><i class="fa fa-check"></i><b>15</b> Overlapping Ranges</a></li>
<li class="chapter" data-level="16" data-path="joining-ranges.html"><a href="joining-ranges.html"><i class="fa fa-check"></i><b>16</b> Joining Ranges</a></li>
<li class="chapter" data-level="17" data-path="finding-the-closest-intervals.html"><a href="finding-the-closest-intervals.html"><i class="fa fa-check"></i><b>17</b> Finding the closest intervals</a></li>
<li class="chapter" data-level="18" data-path="finding-the-k-closest-intervals.html"><a href="finding-the-k-closest-intervals.html"><i class="fa fa-check"></i><b>18</b> Finding the k closest intervals</a></li>
<li class="chapter" data-level="19" data-path="create-count-matrix-from-multiple-pyranges.html"><a href="create-count-matrix-from-multiple-pyranges.html"><i class="fa fa-check"></i><b>19</b> Create count-matrix from multiple PyRanges</a></li>
<li class="chapter" data-level="20" data-path="statistics-colocalization-and-co-occurence-measures.html"><a href="statistics-colocalization-and-co-occurence-measures.html"><i class="fa fa-check"></i><b>20</b> Statistics: colocalization and co-occurence measures</a></li>
<li class="chapter" data-level="21" data-path="statistics-simes-method.html"><a href="statistics-simes-method.html"><i class="fa fa-check"></i><b>21</b> Statistics: Simes method</a></li>
<li class="chapter" data-level="22" data-path="statistics-compute-fisher-exact.html"><a href="statistics-compute-fisher-exact.html"><i class="fa fa-check"></i><b>22</b> Statistics: compute Fisher exact</a></li>
<li class="chapter" data-level="23" data-path="statistics-computing-the-matthews-correlation-coeficcient-between-ranges.html"><a href="statistics-computing-the-matthews-correlation-coeficcient-between-ranges.html"><i class="fa fa-check"></i><b>23</b> Statistics: computing the Matthews correlation coeficcient between ranges</a></li>
<li class="chapter" data-level="24" data-path="statistics-rowbased-operations.html"><a href="statistics-rowbased-operations.html"><i class="fa fa-check"></i><b>24</b> Statistics: rowbased operations</a></li>
<li class="chapter" data-level="25" data-path="turning-ranges-into-rles.html"><a href="turning-ranges-into-rles.html"><i class="fa fa-check"></i><b>25</b> Turning Ranges into RLEs</a></li>
<li class="chapter" data-level="26" data-path="an-introduction-to-rles.html"><a href="an-introduction-to-rles.html"><i class="fa fa-check"></i><b>26</b> An introduction to RLEs</a></li>
<li class="chapter" data-level="27" data-path="run-length-encoding-dicts.html"><a href="run-length-encoding-dicts.html"><i class="fa fa-check"></i><b>27</b> Run Length Encoding dicts</a></li>
<li class="chapter" data-level="28" data-path="subsetting-rles.html"><a href="subsetting-rles.html"><i class="fa fa-check"></i><b>28</b> Subsetting Rles</a></li>
<li class="chapter" data-level="29" data-path="subsetting-pyrles.html"><a href="subsetting-pyrles.html"><i class="fa fa-check"></i><b>29</b> Subsetting PyRles</a></li>
<li class="chapter" data-level="30" data-path="using-multiple-cores.html"><a href="using-multiple-cores.html"><i class="fa fa-check"></i><b>30</b> Using multiple cores</a></li>
<li class="chapter" data-level="31" data-path="genomicfeatures-methods-using-genomic-context.html"><a href="genomicfeatures-methods-using-genomic-context.html"><i class="fa fa-check"></i><b>31</b> GenomicFeatures: methods using genomic context</a></li>
<li class="chapter" data-level="32" data-path="fetching-gene-tracks-as-pyranges.html"><a href="fetching-gene-tracks-as-pyranges.html"><i class="fa fa-check"></i><b>32</b> Fetching gene tracks as PyRanges</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">PyRanges documentation</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="joining-ranges" class="section level1" number="16">
<h1><span class="header-section-number">16</span> Joining Ranges</h1>
<p>You can combine all the intervals that overlap in two PyRanges objects with the join method.
If you do not use a suffix, the default <code>_b</code> is chosen.</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb190-1"><a href="joining-ranges.html#cb190-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyranges <span class="im">as</span> pr</span>
<span id="cb190-2"><a href="joining-ranges.html#cb190-2" aria-hidden="true" tabindex="-1"></a>gr <span class="op">=</span> pr.data.aorta()</span>
<span id="cb190-3"><a href="joining-ranges.html#cb190-3" aria-hidden="true" tabindex="-1"></a>gr2 <span class="op">=</span> pr.data.aorta2()</span>
<span id="cb190-4"><a href="joining-ranges.html#cb190-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gr.join(gr2, suffix<span class="op">=</span><span class="st">&quot;_2&quot;</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+-------+
## | Chromosome   | Start     | End       | Name       | Score     | +6    |
## | (category)   | (int32)   | (int32)   | (object)   | (int64)   | ...   |
## |--------------+-----------+-----------+------------+-----------+-------|
## | chr1         | 9939      | 10138     | H3K27me3   | 7         | ...   |
## | chr1         | 9939      | 10138     | H3K27me3   | 7         | ...   |
## | chr1         | 9939      | 10138     | H3K27me3   | 7         | ...   |
## | chr1         | 9939      | 10138     | H3K27me3   | 7         | ...   |
## | ...          | ...       | ...       | ...        | ...       | ...   |
## | chr1         | 10241     | 10440     | H3K27me3   | 6         | ...   |
## | chr1         | 10241     | 10440     | H3K27me3   | 6         | ...   |
## | chr1         | 10241     | 10440     | H3K27me3   | 6         | ...   |
## | chr1         | 10241     | 10440     | H3K27me3   | 6         | ...   |
## +--------------+-----------+-----------+------------+-----------+-------+
## Stranded PyRanges object has 49 rows and 11 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 6 hidden columns: Strand, Start_2, End_2, Name_2, Score_2, Strand_2</code></pre>
<p>Both methods also take a strandedness option, which can either be <code>"same"</code>, <code>"opposite"</code> or <code>False</code>/<code>None</code></p>
<div class="sourceCode" id="cb192"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb192-1"><a href="joining-ranges.html#cb192-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gr.join(gr2, strandedness<span class="op">=</span><span class="st">&quot;opposite&quot;</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+-------+
## | Chromosome   | Start     | End       | Name       | Score     | +6    |
## | (category)   | (int32)   | (int32)   | (object)   | (int64)   | ...   |
## |--------------+-----------+-----------+------------+-----------+-------|
## | chr1         | 9939      | 10138     | H3K27me3   | 7         | ...   |
## | chr1         | 9939      | 10138     | H3K27me3   | 7         | ...   |
## | chr1         | 9939      | 10138     | H3K27me3   | 7         | ...   |
## | chr1         | 9953      | 10152     | H3K27me3   | 5         | ...   |
## | ...          | ...       | ...       | ...        | ...       | ...   |
## | chr1         | 10127     | 10326     | H3K27me3   | 1         | ...   |
## | chr1         | 10127     | 10326     | H3K27me3   | 1         | ...   |
## | chr1         | 10241     | 10440     | H3K27me3   | 6         | ...   |
## | chr1         | 10241     | 10440     | H3K27me3   | 6         | ...   |
## +--------------+-----------+-----------+------------+-----------+-------+
## Stranded PyRanges object has 22 rows and 11 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 6 hidden columns: Strand, Start_b, End_b, Name_b, Score_b, Strand_b</code></pre>
<p>The join method also takes a how argument, which currently accepts the option
<code>"containment"</code>, <code>"inner"</code> (default), <code>"outer"</code>, <code>"left"</code> and “<code>right</code>”.
Containment requires that the intervals in self be completely within the
intervals in other. The others are similar to SQL-style inner, outer, left and
right joins.</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb194-1"><a href="joining-ranges.html#cb194-1" aria-hidden="true" tabindex="-1"></a>f1 <span class="op">=</span> pr.data.f1()</span>
<span id="cb194-2"><a href="joining-ranges.html#cb194-2" aria-hidden="true" tabindex="-1"></a>f2 <span class="op">=</span> pr.data.f2()</span>
<span id="cb194-3"><a href="joining-ranges.html#cb194-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f2.join(f1, how<span class="op">=</span><span class="st">&quot;containment&quot;</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+-------+
## | Chromosome   |     Start |       End | Name       |     Score | +6    |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |
## |--------------+-----------+-----------+------------+-----------+-------|
## | chr1         |         6 |         7 | b          |         0 | ...   |
## +--------------+-----------+-----------+------------+-----------+-------+
## Stranded PyRanges object has 1 rows and 11 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 6 hidden columns: Strand, Start_b, End_b, Name_b, Score_b, Strand_b</code></pre>
<div class="sourceCode" id="cb196"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb196-1"><a href="joining-ranges.html#cb196-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f1.join(f2, how<span class="op">=</span><span class="st">&quot;outer&quot;</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+-------+
## | Chromosome   |     Start |       End | Name       |     Score | +6    |
## | (category)   |   (int64) |   (int64) | (object)   |   (int64) | ...   |
## |--------------+-----------+-----------+------------+-----------+-------|
## | chr1         |         3 |         6 | interval1  |         0 | ...   |
## | chr1         |         8 |         9 | interval3  |         0 | ...   |
## | chr1         |        -1 |        -1 | -1         |        -1 | ...   |
## | chr1         |        -1 |        -1 | -1         |        -1 | ...   |
## | chr1         |         5 |         7 | interval2  |         0 | ...   |
## | chr1         |        -1 |        -1 | -1         |        -1 | ...   |
## +--------------+-----------+-----------+------------+-----------+-------+
## Stranded PyRanges object has 6 rows and 11 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 6 hidden columns: Strand, Start_b, End_b, Name_b, Score_b, Strand_b
## 
## /usr/share/miniconda/lib/python3.9/site-packages/pyranges-0.0.115-py3.9.egg/pyranges/methods/join.py:87: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
##   scdf = scdf.append(sh)
## /usr/share/miniconda/lib/python3.9/site-packages/pyranges-0.0.115-py3.9.egg/pyranges/methods/join.py:88: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
##   ocdf = ocdf.append(oh)
## /usr/share/miniconda/lib/python3.9/site-packages/pyranges-0.0.115-py3.9.egg/pyranges/methods/join.py:87: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
##   scdf = scdf.append(sh)
## /usr/share/miniconda/lib/python3.9/site-packages/pyranges-0.0.115-py3.9.egg/pyranges/methods/join.py:88: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
##   ocdf = ocdf.append(oh)</code></pre>
<p>After joining, you can use the method new_position() to find the union or
intersection of the joined ranges. By default it uses the two first columns containing “Start” and the two first containing “End”.
Otherwise, the columns to be used can be given by the columns argument.</p>
<div class="sourceCode" id="cb198"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb198-1"><a href="joining-ranges.html#cb198-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f2.join(f1).new_position(<span class="st">&quot;intersection&quot;</span>))</span>
<span id="cb198-2"><a href="joining-ranges.html#cb198-2" aria-hidden="true" tabindex="-1"></a><span class="co"># same as:</span></span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+-------+
## | Chromosome   |     Start |       End | Name       |     Score | +6    |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |
## |--------------+-----------+-----------+------------+-----------+-------|
## | chr1         |         6 |         7 | b          |         0 | ...   |
## +--------------+-----------+-----------+------------+-----------+-------+
## Stranded PyRanges object has 1 rows and 11 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 6 hidden columns: Strand, Start_b, End_b, Name_b, Score_b, Strand_b</code></pre>
<div class="sourceCode" id="cb200"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb200-1"><a href="joining-ranges.html#cb200-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f2.join(f1).new_position(<span class="st">&quot;intersection&quot;</span>, columns<span class="op">=</span>[<span class="st">&quot;Start&quot;</span>, <span class="st">&quot;End&quot;</span>, <span class="st">&quot;Start_b&quot;</span>, <span class="st">&quot;End_b&quot;</span>]))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+-------+
## | Chromosome   |     Start |       End | Name       |     Score | +6    |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |
## |--------------+-----------+-----------+------------+-----------+-------|
## | chr1         |         6 |         7 | b          |         0 | ...   |
## +--------------+-----------+-----------+------------+-----------+-------+
## Stranded PyRanges object has 1 rows and 11 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 6 hidden columns: Strand, Start_b, End_b, Name_b, Score_b, Strand_b</code></pre>
<p>If you want to swap which columns are considered the Start and End, you can use
the swap argument.</p>
<div class="sourceCode" id="cb202"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb202-1"><a href="joining-ranges.html#cb202-1" aria-hidden="true" tabindex="-1"></a>gr1, gr2 <span class="op">=</span> pr.data.chipseq(), pr.data.chipseq_background()</span>
<span id="cb202-2"><a href="joining-ranges.html#cb202-2" aria-hidden="true" tabindex="-1"></a>j <span class="op">=</span> gr1.join(gr2)</span>
<span id="cb202-3"><a href="joining-ranges.html#cb202-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(j)</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+-------+
## | Chromosome   |     Start |       End | Name       |     Score | +6    |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |
## |--------------+-----------+-----------+------------+-----------+-------|
## | chr1         | 226987592 | 226987617 | U0         |         0 | ...   |
## | chr8         |  38747226 |  38747251 | U0         |         0 | ...   |
## | chr15        |  26105515 |  26105540 | U0         |         0 | ...   |
## +--------------+-----------+-----------+------------+-----------+-------+
## Stranded PyRanges object has 3 rows and 11 columns from 3 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 6 hidden columns: Strand, Start_b, End_b, Name_b, Score_b, Strand_b</code></pre>
<div class="sourceCode" id="cb204"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb204-1"><a href="joining-ranges.html#cb204-1" aria-hidden="true" tabindex="-1"></a>j2 <span class="op">=</span> j.new_position(<span class="st">&quot;intersection&quot;</span>)</span>
<span id="cb204-2"><a href="joining-ranges.html#cb204-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(j2)</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+-------+
## | Chromosome   |     Start |       End | Name       |     Score | +6    |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |
## |--------------+-----------+-----------+------------+-----------+-------|
## | chr1         | 226987603 | 226987617 | U0         |         0 | ...   |
## | chr8         |  38747236 |  38747251 | U0         |         0 | ...   |
## | chr15        |  26105515 |  26105518 | U0         |         0 | ...   |
## +--------------+-----------+-----------+------------+-----------+-------+
## Stranded PyRanges object has 3 rows and 11 columns from 3 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 6 hidden columns: Strand, Start_b, End_b, Name_b, Score_b, Strand_b</code></pre>
<div class="sourceCode" id="cb206"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb206-1"><a href="joining-ranges.html#cb206-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(j2.new_position(<span class="st">&quot;swap&quot;</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+-------+
## | Chromosome   |     Start |       End | Name       |     Score | +6    |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |
## |--------------+-----------+-----------+------------+-----------+-------|
## | chr1         | 226987603 | 226987628 | U0         |         0 | ...   |
## | chr8         |  38747236 |  38747261 | U0         |         0 | ...   |
## | chr15        |  26105493 |  26105518 | U0         |         0 | ...   |
## +--------------+-----------+-----------+------------+-----------+-------+
## Stranded PyRanges object has 3 rows and 11 columns from 3 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 6 hidden columns: Strand, Start_b, End_b, Name_b, Score_b, Strand_b</code></pre>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="overlapping-ranges.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="finding-the-closest-intervals.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
