<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Subsetting PyRanges</title>
  <meta name="description" content="Subsetting PyRanges">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Subsetting PyRanges" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Subsetting PyRanges" />
  
  
  

<meta name="author" content="Endre Bakken Stovner">


<meta name="date" content="2022-01-11">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="summarizing-pyranges.html">
<link rel="next" href="applying-custom-methods-to-pairs-of-pyranges.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction to PyRanges</a></li>
<li class="chapter" data-level="2" data-path="loadingcreating-pyranges.html"><a href="loadingcreating-pyranges.html"><i class="fa fa-check"></i><b>2</b> Loading/Creating PyRanges</a></li>
<li class="chapter" data-level="3" data-path="writing-pyranges-to-disk.html"><a href="writing-pyranges-to-disk.html"><i class="fa fa-check"></i><b>3</b> Writing PyRanges to disk</a></li>
<li class="chapter" data-level="4" data-path="subsetting-pyranges.html"><a href="subsetting-pyranges.html"><i class="fa fa-check"></i><b>4</b> Subsetting PyRanges</a></li>
<li class="chapter" data-level="5" data-path="manipulating-the-data-in-pyranges.html"><a href="manipulating-the-data-in-pyranges.html"><i class="fa fa-check"></i><b>5</b> Manipulating the data in PyRanges</a></li>
<li class="chapter" data-level="6" data-path="concatenating-pyranges.html"><a href="concatenating-pyranges.html"><i class="fa fa-check"></i><b>6</b> Concatenating PyRanges</a></li>
<li class="chapter" data-level="7" data-path="piping-pyranges.html"><a href="piping-pyranges.html"><i class="fa fa-check"></i><b>7</b> Piping PyRanges</a></li>
<li class="chapter" data-level="8" data-path="printing-pyranges.html"><a href="printing-pyranges.html"><i class="fa fa-check"></i><b>8</b> Printing PyRanges</a></li>
<li class="chapter" data-level="9" data-path="iterating-over-pyranges.html"><a href="iterating-over-pyranges.html"><i class="fa fa-check"></i><b>9</b> Iterating over PyRanges</a></li>
<li class="chapter" data-level="10" data-path="sorting-pyranges.html"><a href="sorting-pyranges.html"><i class="fa fa-check"></i><b>10</b> Sorting PyRanges</a></li>
<li class="chapter" data-level="11" data-path="summarizing-pyranges.html"><a href="summarizing-pyranges.html"><i class="fa fa-check"></i><b>11</b> Summarizing PyRanges</a></li>
<li class="chapter" data-level="12" data-path="methods-for-manipulating-single-pyranges.html"><a href="methods-for-manipulating-single-pyranges.html"><i class="fa fa-check"></i><b>12</b> Methods for manipulating single PyRanges</a></li>
<li class="chapter" data-level="13" data-path="applying-custom-methods-to-pairs-of-pyranges.html"><a href="applying-custom-methods-to-pairs-of-pyranges.html"><i class="fa fa-check"></i><b>13</b> Applying custom methods to pairs of PyRanges</a></li>
<li class="chapter" data-level="14" data-path="intersecting-ranges.html"><a href="intersecting-ranges.html"><i class="fa fa-check"></i><b>14</b> Intersecting Ranges</a></li>
<li class="chapter" data-level="15" data-path="overlapping-ranges.html"><a href="overlapping-ranges.html"><i class="fa fa-check"></i><b>15</b> Overlapping Ranges</a></li>
<li class="chapter" data-level="16" data-path="joining-ranges.html"><a href="joining-ranges.html"><i class="fa fa-check"></i><b>16</b> Joining Ranges</a></li>
<li class="chapter" data-level="17" data-path="finding-the-closest-intervals.html"><a href="finding-the-closest-intervals.html"><i class="fa fa-check"></i><b>17</b> Finding the closest intervals</a></li>
<li class="chapter" data-level="18" data-path="finding-the-k-closest-intervals.html"><a href="finding-the-k-closest-intervals.html"><i class="fa fa-check"></i><b>18</b> Finding the k closest intervals</a></li>
<li class="chapter" data-level="19" data-path="create-count-matrix-from-multiple-pyranges.html"><a href="create-count-matrix-from-multiple-pyranges.html"><i class="fa fa-check"></i><b>19</b> Create count-matrix from multiple PyRanges</a></li>
<li class="chapter" data-level="20" data-path="statistics-colocalization-and-co-occurence-measures.html"><a href="statistics-colocalization-and-co-occurence-measures.html"><i class="fa fa-check"></i><b>20</b> Statistics: colocalization and co-occurence measures</a></li>
<li class="chapter" data-level="21" data-path="statistics-simes-method.html"><a href="statistics-simes-method.html"><i class="fa fa-check"></i><b>21</b> Statistics: Simes method</a></li>
<li class="chapter" data-level="22" data-path="statistics-compute-fisher-exact.html"><a href="statistics-compute-fisher-exact.html"><i class="fa fa-check"></i><b>22</b> Statistics: compute Fisher exact</a></li>
<li class="chapter" data-level="23" data-path="statistics-computing-the-matthews-correlation-coeficcient-between-ranges.html"><a href="statistics-computing-the-matthews-correlation-coeficcient-between-ranges.html"><i class="fa fa-check"></i><b>23</b> Statistics: computing the Matthews correlation coeficcient between ranges</a></li>
<li class="chapter" data-level="24" data-path="statistics-rowbased-operations.html"><a href="statistics-rowbased-operations.html"><i class="fa fa-check"></i><b>24</b> Statistics: rowbased operations</a></li>
<li class="chapter" data-level="25" data-path="turning-ranges-into-rles.html"><a href="turning-ranges-into-rles.html"><i class="fa fa-check"></i><b>25</b> Turning Ranges into RLEs</a></li>
<li class="chapter" data-level="26" data-path="an-introduction-to-rles.html"><a href="an-introduction-to-rles.html"><i class="fa fa-check"></i><b>26</b> An introduction to RLEs</a></li>
<li class="chapter" data-level="27" data-path="run-length-encoding-dicts.html"><a href="run-length-encoding-dicts.html"><i class="fa fa-check"></i><b>27</b> Run Length Encoding dicts</a></li>
<li class="chapter" data-level="28" data-path="subsetting-rles.html"><a href="subsetting-rles.html"><i class="fa fa-check"></i><b>28</b> Subsetting Rles</a></li>
<li class="chapter" data-level="29" data-path="subsetting-pyrles.html"><a href="subsetting-pyrles.html"><i class="fa fa-check"></i><b>29</b> Subsetting PyRles</a></li>
<li class="chapter" data-level="30" data-path="using-multiple-cores.html"><a href="using-multiple-cores.html"><i class="fa fa-check"></i><b>30</b> Using multiple cores</a></li>
<li class="chapter" data-level="31" data-path="genomicfeatures-methods-using-genomic-context.html"><a href="genomicfeatures-methods-using-genomic-context.html"><i class="fa fa-check"></i><b>31</b> GenomicFeatures: methods using genomic context</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Subsetting PyRanges</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="methods-for-manipulating-single-pyranges" class="section level1" number="12">
<h1><span class="header-section-number">12</span> Methods for manipulating single PyRanges</h1>
<p>There are several methods for manipulating the contents of a PyRanges.</p>
<p><code>merge</code> creates a union of all the intervals in the ranges:</p>
<p>The merge-method also takes the flag count to let you know the number of intervals that were merged.</p>
<div class="sourceCode" id="cb134"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb134-1"><a href="methods-for-manipulating-single-pyranges.html#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyranges <span class="im">as</span> pr</span>
<span id="cb134-2"><a href="methods-for-manipulating-single-pyranges.html#cb134-2" aria-hidden="true" tabindex="-1"></a>f1 <span class="op">=</span> pr.data.f1()</span>
<span id="cb134-3"><a href="methods-for-manipulating-single-pyranges.html#cb134-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f1.merge(count<span class="op">=</span><span class="va">True</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+--------------+-----------+
## | Chromosome   |     Start |       End | Strand       |     Count |
## | (category)   |   (int32) |   (int32) | (category)   |   (int32) |
## |--------------+-----------+-----------+--------------+-----------|
## | chr1         |         3 |         6 | +            |         1 |
## | chr1         |         8 |         9 | +            |         1 |
## | chr1         |         5 |         7 | -            |         1 |
## +--------------+-----------+-----------+--------------+-----------+
## Stranded PyRanges object has 3 rows and 5 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.</code></pre>
<p>Cluster finds which intervals overlap, but gives each interval a cluster ID instead of merging them:</p>
<div class="sourceCode" id="cb136"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb136-1"><a href="methods-for-manipulating-single-pyranges.html#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyranges <span class="im">as</span> pr</span>
<span id="cb136-2"><a href="methods-for-manipulating-single-pyranges.html#cb136-2" aria-hidden="true" tabindex="-1"></a>f1 <span class="op">=</span> pr.data.f1()</span>
<span id="cb136-3"><a href="methods-for-manipulating-single-pyranges.html#cb136-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f1.cluster())</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+-------+
## | Chromosome   |     Start |       End | Name       |     Score | +2    |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |
## |--------------+-----------+-----------+------------+-----------+-------|
## | chr1         |         3 |         6 | interval1  |         0 | ...   |
## | chr1         |         8 |         9 | interval3  |         0 | ...   |
## | chr1         |         5 |         7 | interval2  |         0 | ...   |
## +--------------+-----------+-----------+------------+-----------+-------+
## Stranded PyRanges object has 3 rows and 7 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 2 hidden columns: Strand, Cluster</code></pre>
<div class="sourceCode" id="cb138"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb138-1"><a href="methods-for-manipulating-single-pyranges.html#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f1.cluster(strand<span class="op">=</span><span class="va">True</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+-------+
## | Chromosome   |     Start |       End | Name       |     Score | +2    |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |
## |--------------+-----------+-----------+------------+-----------+-------|
## | chr1         |         3 |         6 | interval1  |         0 | ...   |
## | chr1         |         8 |         9 | interval3  |         0 | ...   |
## | chr1         |         5 |         7 | interval2  |         0 | ...   |
## +--------------+-----------+-----------+------------+-----------+-------+
## Stranded PyRanges object has 3 rows and 7 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 2 hidden columns: Strand, Cluster</code></pre>
<div class="sourceCode" id="cb140"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb140-1"><a href="methods-for-manipulating-single-pyranges.html#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f1.cluster(slack<span class="op">=</span><span class="dv">2</span>, strand<span class="op">=</span><span class="va">True</span>, count<span class="op">=</span><span class="va">True</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+-------+
## | Chromosome   |     Start |       End | Name       |     Score | +3    |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | ...   |
## |--------------+-----------+-----------+------------+-----------+-------|
## | chr1         |         3 |         6 | interval1  |         0 | ...   |
## | chr1         |         8 |         9 | interval3  |         0 | ...   |
## | chr1         |         5 |         7 | interval2  |         0 | ...   |
## +--------------+-----------+-----------+------------+-----------+-------+
## Stranded PyRanges object has 3 rows and 8 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 3 hidden columns: Strand, Cluster, Count</code></pre>
<p>It also takes an flag count.</p>
<p>Both cluster and merge take the argument slack, so that you can merge features
which are not directly overlapping. If you set slack to -1 you avoid merging bookended features.
To only merge features overlapping by at least X, set slack to -X.</p>
<p>To cluster or merge only those rows which have the same values for a certain column, you can use the argument by to cluster and merge.</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb142-1"><a href="methods-for-manipulating-single-pyranges.html#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyranges <span class="im">as</span> pr</span>
<span id="cb142-2"><a href="methods-for-manipulating-single-pyranges.html#cb142-2" aria-hidden="true" tabindex="-1"></a>gr <span class="op">=</span> pr.data.ensembl_gtf()</span>
<span id="cb142-3"><a href="methods-for-manipulating-single-pyranges.html#cb142-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gr)</span></code></pre></div>
<pre><code>## +--------------+------------+--------------+-----------+-----------+-------+
## | Chromosome   | Source     | Feature      | Start     | End       | +23   |
## | (category)   | (object)   | (category)   | (int32)   | (int32)   | ...   |
## |--------------+------------+--------------+-----------+-----------+-------|
## | 1            | havana     | gene         | 11868     | 14409     | ...   |
## | 1            | havana     | transcript   | 11868     | 14409     | ...   |
## | 1            | havana     | exon         | 11868     | 12227     | ...   |
## | 1            | havana     | exon         | 12612     | 12721     | ...   |
## | ...          | ...        | ...          | ...       | ...       | ...   |
## | 1            | havana     | gene         | 1173055   | 1179555   | ...   |
## | 1            | havana     | transcript   | 1173055   | 1179555   | ...   |
## | 1            | havana     | exon         | 1179364   | 1179555   | ...   |
## | 1            | havana     | exon         | 1173055   | 1176396   | ...   |
## +--------------+------------+--------------+-----------+-----------+-------+
## Stranded PyRanges object has 2,446 rows and 28 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 23 hidden columns: Score, Strand, Frame, gene_biotype, gene_id, gene_name, gene_source, ... (+ 16 more.)</code></pre>
<div class="sourceCode" id="cb144"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb144-1"><a href="methods-for-manipulating-single-pyranges.html#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(gr.cluster(by<span class="op">=</span><span class="st">&quot;gene_id&quot;</span>))</span></code></pre></div>
<pre><code>## +--------------+----------------+----------------+-----------+-------+
## | Chromosome   | Source         | Feature        | Start     | +25   |
## | (category)   | (object)       | (category)     | (int32)   | ...   |
## |--------------+----------------+----------------+-----------+-------|
## | 1            | ensembl_havana | gene           | 1173883   | ...   |
## | 1            | havana         | transcript     | 1173883   | ...   |
## | 1            | havana         | exon           | 1173883   | ...   |
## | 1            | havana         | transcript     | 1173902   | ...   |
## | ...          | ...            | ...            | ...       | ...   |
## | 1            | ensembl_havana | stop_codon     | 450739    | ...   |
## | 1            | ensembl_havana | CDS            | 450742    | ...   |
## | 1            | ensembl_havana | start_codon    | 451675    | ...   |
## | 1            | ensembl_havana | five_prime_utr | 451678    | ...   |
## +--------------+----------------+----------------+-----------+-------+
## Stranded PyRanges object has 2,446 rows and 29 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 25 hidden columns: End, Score, Strand, Frame, gene_biotype, gene_id, gene_name, ... (+ 18 more.)</code></pre>
<p>If you want to split a pyrange on the overlapping intervals, you can use split:</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb146-1"><a href="methods-for-manipulating-single-pyranges.html#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyranges <span class="im">as</span> pr</span>
<span id="cb146-2"><a href="methods-for-manipulating-single-pyranges.html#cb146-2" aria-hidden="true" tabindex="-1"></a>f1 <span class="op">=</span> pr.data.f1()</span>
<span id="cb146-3"><a href="methods-for-manipulating-single-pyranges.html#cb146-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f1)</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+--------------+
## | Chromosome   |     Start |       End | Name       |     Score | Strand       |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |
## |--------------+-----------+-----------+------------+-----------+--------------|
## | chr1         |         3 |         6 | interval1  |         0 | +            |
## | chr1         |         8 |         9 | interval3  |         0 | +            |
## | chr1         |         5 |         7 | interval2  |         0 | -            |
## +--------------+-----------+-----------+------------+-----------+--------------+
## Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.</code></pre>
<div class="sourceCode" id="cb148"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb148-1"><a href="methods-for-manipulating-single-pyranges.html#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f1.split())</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+
## | Chromosome   |     Start |       End | Strand     |
## | (object)     |   (int32) |   (int32) | (object)   |
## |--------------+-----------+-----------+------------|
## | chr1         |         3 |         6 | +          |
## | chr1         |         8 |         9 | +          |
## | chr1         |         5 |         7 | -          |
## +--------------+-----------+-----------+------------+
## Stranded PyRanges object has 3 rows and 4 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.</code></pre>
<p><code>five_end</code> finds the starts of the regions (taking direction of transcription
into account).</p>
<div class="sourceCode" id="cb150"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb150-1"><a href="methods-for-manipulating-single-pyranges.html#cb150-1" aria-hidden="true" tabindex="-1"></a>f1.five_end()</span>
<span id="cb150-2"><a href="methods-for-manipulating-single-pyranges.html#cb150-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f1.five_end())</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+--------------+
## | Chromosome   |     Start |       End | Name       |     Score | Strand       |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |
## |--------------+-----------+-----------+------------+-----------+--------------|
## | chr1         |         3 |         4 | interval1  |         0 | +            |
## | chr1         |         8 |         9 | interval3  |         0 | +            |
## | chr1         |         6 |         7 | interval2  |         0 | -            |
## +--------------+-----------+-----------+------------+-----------+--------------+
## Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.</code></pre>
<p><code>three_end</code> finds the ends of the regions (taking direction of transcription into account).</p>
<div class="sourceCode" id="cb152"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb152-1"><a href="methods-for-manipulating-single-pyranges.html#cb152-1" aria-hidden="true" tabindex="-1"></a>f1.three_end()</span>
<span id="cb152-2"><a href="methods-for-manipulating-single-pyranges.html#cb152-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f1.three_end())</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+--------------+
## | Chromosome   |     Start |       End | Name       |     Score | Strand       |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |
## |--------------+-----------+-----------+------------+-----------+--------------|
## | chr1         |         5 |         6 | interval1  |         0 | +            |
## | chr1         |         8 |         9 | interval3  |         0 | +            |
## | chr1         |         5 |         6 | interval2  |         0 | -            |
## +--------------+-----------+-----------+------------+-----------+--------------+
## Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.</code></pre>
<p><code>extend</code> (also aliased <code>slack</code>) extends the starts and ends of your interval:</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb154-1"><a href="methods-for-manipulating-single-pyranges.html#cb154-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f1.slack(<span class="dv">5</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+--------------+
## | Chromosome   |     Start |       End | Name       |     Score | Strand       |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |
## |--------------+-----------+-----------+------------+-----------+--------------|
## | chr1         |         0 |        11 | interval1  |         0 | +            |
## | chr1         |         3 |        14 | interval3  |         0 | +            |
## | chr1         |         0 |        12 | interval2  |         0 | -            |
## +--------------+-----------+-----------+------------+-----------+--------------+
## Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.</code></pre>
<div class="sourceCode" id="cb156"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb156-1"><a href="methods-for-manipulating-single-pyranges.html#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(f1.slack({<span class="st">&quot;5&quot;</span>: <span class="dv">2</span>, <span class="st">&quot;3&quot;</span>: <span class="op">-</span><span class="dv">1</span>}))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+--------------+
## | Chromosome   |     Start |       End | Name       |     Score | Strand       |
## | (category)   |   (int32) |   (int32) | (object)   |   (int64) | (category)   |
## |--------------+-----------+-----------+------------+-----------+--------------|
## | chr1         |         1 |         5 | interval1  |         0 | +            |
## | chr1         |         6 |         8 | interval3  |         0 | +            |
## | chr1         |         6 |         9 | interval2  |         0 | -            |
## +--------------+-----------+-----------+------------+-----------+--------------+
## Stranded PyRanges object has 3 rows and 6 columns from 1 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.</code></pre>
<p><code>window</code> splits your data into windows:</p>
<div class="sourceCode" id="cb158"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb158-1"><a href="methods-for-manipulating-single-pyranges.html#cb158-1" aria-hidden="true" tabindex="-1"></a>exons <span class="op">=</span> pr.data.exons()</span>
<span id="cb158-2"><a href="methods-for-manipulating-single-pyranges.html#cb158-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(exons)</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+-------+
## | Chromosome   | Start     | End       | +3    |
## | (category)   | (int32)   | (int32)   | ...   |
## |--------------+-----------+-----------+-------|
## | chrX         | 135721701 | 135721963 | ...   |
## | chrX         | 135574120 | 135574598 | ...   |
## | chrX         | 47868945  | 47869126  | ...   |
## | chrX         | 77294333  | 77294480  | ...   |
## | ...          | ...       | ...       | ...   |
## | chrY         | 15409586  | 15409728  | ...   |
## | chrY         | 15478146  | 15478273  | ...   |
## | chrY         | 15360258  | 15361762  | ...   |
## | chrY         | 15467254  | 15467278  | ...   |
## +--------------+-----------+-----------+-------+
## Stranded PyRanges object has 1,000 rows and 6 columns from 2 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 3 hidden columns: Name, Score, Strand</code></pre>
<div class="sourceCode" id="cb160"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb160-1"><a href="methods-for-manipulating-single-pyranges.html#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(exons.window(<span class="dv">5</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+-------+
## | Chromosome   | Start     | End       | +3    |
## | (category)   | (int32)   | (int32)   | ...   |
## |--------------+-----------+-----------+-------|
## | chrX         | 135721701 | 135721706 | ...   |
## | chrX         | 135721706 | 135721711 | ...   |
## | chrX         | 135721711 | 135721716 | ...   |
## | chrX         | 135721716 | 135721721 | ...   |
## | ...          | ...       | ...       | ...   |
## | chrY         | 15467259  | 15467264  | ...   |
## | chrY         | 15467264  | 15467269  | ...   |
## | chrY         | 15467269  | 15467274  | ...   |
## | chrY         | 15467274  | 15467278  | ...   |
## +--------------+-----------+-----------+-------+
## Stranded PyRanges object has 61,268 rows and 6 columns from 2 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 3 hidden columns: Name, Score, Strand</code></pre>
<p>If you want to all windows to start at a multiple of window_size, use the tile function:</p>
<p>You can use the overlap flag to see how many basepairs each tile overlapped with
the underlying interval.</p>
<div class="sourceCode" id="cb162"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb162-1"><a href="methods-for-manipulating-single-pyranges.html#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(exons.tile(<span class="dv">5</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+-------+
## | Chromosome   | Start     | End       | +3    |
## | (category)   | (int32)   | (int32)   | ...   |
## |--------------+-----------+-----------+-------|
## | chrX         | 135721700 | 135721705 | ...   |
## | chrX         | 135721705 | 135721710 | ...   |
## | chrX         | 135721710 | 135721715 | ...   |
## | chrX         | 135721715 | 135721720 | ...   |
## | ...          | ...       | ...       | ...   |
## | chrY         | 15467260  | 15467265  | ...   |
## | chrY         | 15467265  | 15467270  | ...   |
## | chrY         | 15467270  | 15467275  | ...   |
## | chrY         | 15467275  | 15467280  | ...   |
## +--------------+-----------+-----------+-------+
## Stranded PyRanges object has 61,643 rows and 6 columns from 2 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 3 hidden columns: Name, Score, Strand</code></pre>
<div class="sourceCode" id="cb164"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb164-1"><a href="methods-for-manipulating-single-pyranges.html#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(exons.tile(<span class="dv">5</span>, overlap<span class="op">=</span><span class="va">True</span>))</span></code></pre></div>
<pre><code>## +--------------+-----------+-----------+-------+
## | Chromosome   | Start     | End       | +4    |
## | (category)   | (int32)   | (int32)   | ...   |
## |--------------+-----------+-----------+-------|
## | chrX         | 135721700 | 135721705 | ...   |
## | chrX         | 135721705 | 135721710 | ...   |
## | chrX         | 135721710 | 135721715 | ...   |
## | chrX         | 135721715 | 135721720 | ...   |
## | ...          | ...       | ...       | ...   |
## | chrY         | 15467260  | 15467265  | ...   |
## | chrY         | 15467265  | 15467270  | ...   |
## | chrY         | 15467270  | 15467275  | ...   |
## | chrY         | 15467275  | 15467280  | ...   |
## +--------------+-----------+-----------+-------+
## Stranded PyRanges object has 61,643 rows and 7 columns from 2 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## 4 hidden columns: Name, Score, Strand, TileOverlap</code></pre>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="summarizing-pyranges.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="applying-custom-methods-to-pairs-of-pyranges.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
